runner.py

```diff
@@ -0,0 +1,707 @@
+    SYSPAGE = {
+        'ia32-generic': [
+            'uart16550',
+            'pc-ata',
+            'psh'
+        ],
+        'riscv64-virt': [
+            'uart16550',
+            'dummyfs',
+            'psh'
+        ]
+    }
+
+    def __init__(self, target):
+        if target not in TargetBuilder.TARGETS:
+            raise ValueError(f"invalid target: {target}")
+
+        self.env = os.environ.copy()
+        self.target = target
+        self.fs_path = PRTOS_PROJECT_DIR / f"_fs/{self.target}"
+        self.env['TARGET'] = self.target
+        self.env['CONSOLE'] = 'serial'
+
+    def __str__(self):
+        return self.target
+
+    def fs_mkdir(self, path):
+        if path.anchor == '/':
+            path = path.relative_to('/')
+        abs_path = self.fs_path / path
+        abs_path.mkdir(exist_ok=True)
+
+    def fs_install(self, path, file, mode):
+        if path.anchor == '/':
+            path = path.relative_to('/')
+        abs_path = self.fs_path / path
+        shutil.copy(file, abs_path)
+        abs_path.joinpath(file.name).chmod(mode)
+
+    def prebuild_action(self, syspage=None, **kwargs):
+        pass
+
+    def postbuild_action(self, syspage=None, **kwargs):
+        pass
+
+    def run_command(self, args, handle_error=True):
+        proc = subprocess.Popen(
+            args,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env=self.env,
+            universal_newlines=True,
+            cwd=PRTOS_PROJECT_DIR
+        )
+
+        out, err = proc.communicate()
+        if proc.returncode != 0 and handle_error:
+            logging.error(f"Command {' '.join(args)} for {self.target} failed!")
+            logging.error(out)
+            logging.error(err)
+            # Fail at build stage should not happen, exit immediately
+            sys.exit(1)
+        elif proc.returncode != 0:
+            logging.debug(f"Command {' '.join(args)} for {self.target} failed!")
+        else:
+            logging.debug(f"Command {' '.join(args)} for {self.target} success!")
+
+        return proc.returncode, out, err
+
+    def build(self, syspage=None):
+        if syspage:
+            syspage_with_defaults = list(set(syspage + TargetBuilder.SYSPAGE[self.target]))
+            self.env['SYSPAGE'] = " ".join(syspage_with_defaults)
+
+        self.prebuild_action(syspage)
+
+        logging.info(f"Building {self.env['TARGET']} with syspage: {self.env['SYSPAGE']}")
+        self.run_command(['./docker-build.sh', 'clean', 'core', 'fs', 'image', 'project'])
+
+        self.postbuild_action(syspage)
+
+
+class IA32Builder(TargetBuilder):
+    def __init__(self):
+        super().__init__('ia32-generic')
+
+    def prebuild_action(self, syspage=None, **kwargs):
+        # We run syspage progs from filesystem, set only default syspage
+        self.env['SYSPAGE'] = " ".join(self.SYSPAGE[self.target])
+
+    def postbuild_action(self, syspage=None, **kwargs):
```

Comment:
invalid naming - `syspage`

Reply:
I left here `user_syspage`. We use `syspage` word in YAML config so it also have to be changed to `exec` or something. On the other hand only `ia32-generic` arch has fs system for now so on the other architectures it will be embedded for real into syspage. Let me know what you think.

In other words we have to distinguish:
1. Binaries embedded into syspage
2. Binaries embedded into syspage that are executed on OS start (like drivers)
3. Binaries embedded into filesystem

Reply:
Linked to my previous comment - the `DeviceRunner` should provide the abstraction over "uploading files" (putting them on target) and executing programs. For sure we shouldn't use `syspage` naming, as for most targets we will not be using syspage for test files.

I assume building base image with all necessary drivers to run on a particular target is the responsibility of the build system, not a responsibility of this script, so we don't need to know set of necessary drivers for each of the targets to be workable.

Right now regarding different targets we can upload files by:
 - putting them in the rootfs image (`ia32-generic` (ext2), `armv7a7-imx6ull` (jffs), `riscv64-*` in the near future)
 - uploading files via `plo` - phfs on serial / USB CDC (`armv7a9-zynq7000`, `armv7m7-imxrt106x`, `armv7m7-imxrt117x`)
 - `armv7m4-stm32l4x6` - currently only by appending directly to syspage (AFAIK)